moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp

```diff
@@ -276,21 +276,24 @@ bool planning_scene_monitor::CurrentStateMonitor::waitForCurrentState(const ros:
   ros::WallTime start = ros::WallTime::now();
   ros::WallDuration elapsed(0, 0);
   ros::WallDuration timeout(wait_time);
+  ros::Time start_ = ros::Time::now();
+  ros::Duration elapsed_(0, 0);
 
-  boost::mutex::scoped_lock lock(state_update_lock_);
-  while (current_state_time_ < t)
+  while (elapsed < timeout)
   {
-    state_update_condition_.wait_for(lock, boost::chrono::nanoseconds((timeout - elapsed).toNSec()));
-    elapsed = ros::WallTime::now() - start;
-    if (elapsed > timeout)
     {
-      ROS_INFO_STREAM("Didn't received robot state (joint angles) with recent timestamp within "
-                      << wait_time << " seconds.\n"
-                      << "Check clock synchronization if your are running ROS across multiple machines!");
-      return false;
+      boost::mutex::scoped_lock lock(state_update_lock_);
+      state_update_condition_.wait_for(lock, boost::chrono::nanoseconds((timeout - elapsed).toNSec()));
     }
+    elapsed = ros::WallTime::now() - start;
+    elapsed_ = ros::Time::now() - start_;
+    if (haveCompleteState(elapsed_))
```

Comment:
`haveCompleteState(age)` returns true if all joints have had an update younger than `age`, so I think `elapsed_` is the right parameter to hand in?

If your suggested overload would return true if all joints are newer than a certain time, that sounds fine too. It would certainly look better in this case.

Reply:
> so I think elapsed_ is the right parameter to hand in?

No, because it's based on `start_d` which is initialized via `ros::Time::now()`. But `waitForCurrentState` is passed a `t` parameter that defines "current" already.

> If your suggested overload would return true if all joints are newer than a certain time, that sounds fine too. It would certainly look better in this case.

I believe not just in this case. During execution of `haveCompleteState` a negligible amount of time passes, so you can't use it to accurately to wait for a time stamp (it might wait a tiny bit longer).
Also, the only place where `age` is used is in [`ros::Time old = now - age`](https://github.com/ros-planning/moveit/blob/master/moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp#L222).
It's just stupid to convert from time stamp to duration to time stamp while time passes.